types/styled-components/index.d.ts

```diff
@@ -65,7 +66,7 @@ export type StyledComponentProps<
         A
     > & Partial<Pick<React.ComponentPropsWithRef<C> & O, A>>,
     T
->;
+> & { children?: React.ReactNode };
```

Comment:
this is the fix for https://github.com/DefinitelyTyped/DefinitelyTyped/issues/31491. Essentially, we just always tack on `children` to whatever props are provided (since this is what React does as well in its typings).

Reply:
You can specifically type a component to not accept children if you avoid the react types, or you can specify _exactly_ the kind of child that you want such as for render callbacks.

```jsx
<Component>
  {({ text }) => (
    <p>{text}</p>
  )}
</Component>
```

For more information see #31945.

Reply:
The problem is that that works by accident. "since this is what React does as well in its typings" is a bad example (not your fault, it's just a terrible precedent that's still standing because of backwards compatibility).

Reply:
The fix I was going to suggest is to update `React.ComponentType` to be based off `JSXElementConstructor` instead of `ComponentType` but that has already been done on that side. The change to use `JSXElementConstructor` probably just needs to be propagated to `ReactType`.

Reply:
Thanks for the insightful feedback both @Lavoaster and @Jessidhia! I had missed seeing #31945 previously and didn't understand the full context of the problem.

I'm hoping my new commit I just pushed is a reasonable solution to the issue.

@Jessidhia as you mentioned, `type React.ComponentProps` was recently changed to use `extends ((props: infer P) => ReactNode) | (new (props: infer P) => Component<any>))`, which means things are already working well for functional components. However, `React.Component` only declares `children` as part of its `props` type and not in its constructor:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/0b417416fe69ce507a5224f32924da296667b1fe/types/react/index.d.ts#L425

I believe this means that we won't be able to make `(new (props: infer P) => Component<any>)` useful to us. So instead, my new changes add `{ children?: React.ReactNode }` to `StyledComponentProps` only if our base component type is a `React.ComponentClass`. I _think_ this'll give us `children` where appropriate without being overly permissive where we shouldn't be.

---

Just to give context on why I'm working on this fix as part of this PR, this fix ends up being really important for react-native workflows (since all elements you use are components). Without the fix, you can't even do really common and basic workflows, like the following:

```ts
const element = <StyledView>
  <StyledText>Hello World!</StyledText>
</StyledView>;
```